src/pages/Animal/Feed/index.tsx

```diff
@@ -1,40 +1,128 @@
 
-  const [fetchedPets, setFetchedPets] = useState<FirebaseFirestoreTypes.DocumentData[]>([]);
-
   useLayoutEffect(() => {
     setStatusBarBackgroundColor(Theme.elements.statusBarSecondaryDark, false);
   }, [navigation]);
 
-  const fetchPets = (): void => {
-    animalAPI.getAll()
-      .then((result) => {
-        const data = result.docs.map((doc) => ({ id: uuidv4(), ...(doc.data()) }));
-        setFetchedPets(data);
+  const animalKey = (animalItem: Animal): string => animalItem.id;
+
+  const formatAnimal = (pet: Animal): JSX.Element => (
+    <AnimalCard
+      key={pet?.id}
+      imageUrl={`${pet?.pictures.length > 0 ? `${pet?.pictures[0]}` : null}`}
+      body={(
+        <CardTextContainer>
+          <CardTextRow>
+            <CardText>{SexNames[pet?.sex as Sex]}</CardText>
+            <CardText>{SizeNames[pet?.size as Size]}</CardText>
+            <CardText>{AgeNames[pet?.age as Age]}</CardText>
+          </CardTextRow>
+          <CardTextRow>
+            <CardText>
+              {formatLocation(pet?.ownerDocument)}
+            </CardText>
+          </CardTextRow>
+        </CardTextContainer>
+      )}
+      title={pet?.name}
+      headerOptions={(
+        <MaterialCommunityIcons
+          name="heart-outline"
+          size={24}
+          color={Theme.elements.headerText}
+        />
+      )}
+      headerBackground={Theme.elements.headerSecondary}
+      onPress={() => navigation.navigate('AnimalDetails', { animalUID: pet?.id })}
+    />
+  );
+
+  // Fetch pets function declaration
+  const fetchPets = (
+    lastElement: Animal | null, pageNumber: number, pageSize: number,
+  ): Promise<Array<Animal>> => {
+    let query : FirebaseFirestoreTypes.Query;
+    const orderBy = 'name';
+
+    if (pageNumber === 1 || lastElement == null) {
+      query = animalAPI.createQuery({
+        limit: pageSize,
+        orderBy,
       });
-  };
+    } else {
+      query = animalAPI.createQuery({
+        limit: (pageNumber - 1) * pageSize,
+        orderBy,
+        startAfter: lastElement.name,
+      });
+    }
 
-  useEffect(() => fetchPets(), []);
+    return new Promise((resolve, reject) => {
+      query.get()
+        .then((response) => {
+          const promisesArray : Array<Promise<Animal>> = [];
+
+          response.forEach((childSnapshot) => promisesArray.push(new Promise((resolveItem, rejectItem) => {
+            const item = childSnapshot.data();
+            item.id = childSnapshot.id;
+            if (item?.owner !== undefined) {
+              userAPI.getReference(item?.owner)
+                .then((userDocument) => {
+                  item.ownerDocument = userDocument.data();
+                  resolveItem(item as Animal);
+                })
+                .catch((error) => rejectItem(error));
+            } else {
+              resolveItem(item as Animal);
+            }
+          })));
+
+          Promise.all(promisesArray).then((animalArray) => resolve(animalArray));
```

Comment:
Don't forget to write a catch for this Promise.all.

Reply:
Solved